app/scripts/lib/account-tracker.js

```diff
@@ -158,90 +280,148 @@ export default class AccountTracker {
+      addresses.forEach((address) => {
+        delete accountsByChainId[chainId][address];
+      });
+    });
     // save accounts state
-    this.store.updateState({ accounts });
+    this.store.updateState({ accounts, accountsByChainId });
   }
 
   /**
    * Removes all addresses and associated balances
    */
 
   clearAccounts() {
-    this.store.updateState({ accounts: {} });
+    const { accountsByChainId } = this.store.getState();
+    Object.keys(accountsByChainId).forEach((chainId) => {
+      accountsByChainId[chainId] = {};
+    });
+    this.store.updateState({ accounts: {}, accountsByChainId });
   }
 
   /**
-   * Given a block, updates this AccountTracker's currentBlockGasLimit, and then updates each local account's balance
-   * via EthQuery
+   * Given a block, updates this AccountTracker's currentBlockGasLimit and currentBlockGasLimitByChainId and then updates
+   * each local account's balance via EthQuery
    *
    * @private
    * @param {number} blockNumber - the block number to update to.
    * @fires 'block' The updated state, if all account updates are successful
    */
   async _updateForBlock(blockNumber) {
-    this._currentBlockNumber = blockNumber;
+    await this._updateForBlockByNetworkClientId(null, blockNumber);
+  }
+
+  /**
+   * Given a block, updates this AccountTracker's currentBlockGasLimitByChainId, and then updates each local account's balance
+   * via EthQuery
+   *
+   * @private
+   * @param {string} networkClientId - optional network client ID to use instead of the globally selected network.
+   * @param {number} blockNumber - the block number to update to.
+   * @fires 'block' The updated state, if all account updates are successful
+   */
+  async _updateForBlockByNetworkClientId(networkClientId, blockNumber) {
+    const { chainId, provider } =
+      this.#getCorrectNetworkClient(networkClientId);
+    this._currentBlockNumberByChainId[chainId] = blockNumber;
 
     // block gasLimit polling shouldn't be in account-tracker shouldn't be here...
-    const currentBlock = await this._query.getBlockByNumber(blockNumber, false);
+    const currentBlock = await pify(new EthQuery(provider)).getBlockByNumber(
+      blockNumber,
+      false,
+    );
     if (!currentBlock) {
       return;
     }
     const currentBlockGasLimit = currentBlock.gasLimit;
-    this.store.updateState({ currentBlockGasLimit });
+    const { currentBlockGasLimitByChainId } = this.store.getState();
+    currentBlockGasLimitByChainId[chainId] = currentBlockGasLimit;
+    this.store.updateState({ currentBlockGasLimitByChainId });
+    if (chainId === this.getCurrentChainId()) {
+      this.store.updateState({
+        currentBlockGasLimit,
+      });
+    }
 
     try {
-      await this._updateAccounts();
+      await this._updateAccountsWithNetworkClientId(networkClientId);
     } catch (err) {
       log.error(err);
     }
   }
 
+  async _updateAccounts() {
+    this._updateAccountsWithNetworkClientId();
```

Comment:
This seems a bit strange to me. We call the `_updateAccountsWithNetworkClientId` in the generic `_updateAcounts` but without the networkClientId... Can we make this called updateAllAccounts since its looping through all "active" networkClientIds. And then hmmm it feels weird to me that `networkClientId` is an optional arg to `_updateAccountsWithNetworkClientId`. 🤔

Reply:
I think `_updateAccounts` -> `_updateAllAccounts` and `_updateAccountsWithNetworkClientId` -> `_updateAccounts` ?

Reply:
Oh hmm and then there's a singular `updateAccount`. Haha. I see that these are all private methods so its kindof nitpicky but its kindof confusing.

Reply:
Ok separate question here. Is this call not redundant? Won't it happen again on line 406? Or is there some chance that the "active chainId" doesn't already have a polling set.

Reply:
> I think _updateAccounts -> _updateAccountsAllActiveNetworks and _updateAccountsSingleNetwork -> _updateAccounts ?

Looking at this again I actually think that `_updateAccounts` should only be meant to target the globally selected network since it's called in `NetworkController:networkDidChange` and `_loginUser` in MetaMaskController. 

However, we still do need a method that updates accounts for all active networks (that is triggered in the `this.onboardingController.store.subscribe` and `this.preferencesController.store.subscribe` of this controller). So let's add a separate `_updateAccountsAllActiveNetworks` to handle those cases.

To summarize:
* `_updateAccountsWithNetworkClientId` is renamed to `_updateAccounts`
* `_updateAccounts` is renamed to `_updateAccountsAllActiveNetworks`


> Oh hmm and then there's a singular updateAccount. Haha. I see that these are all private methods so its kindof nitpicky but its kindof confusing.

Agreed. Was confusing to me when making changes as well.

> Ok separate question here. Is this call not redundant? Won't it happen again on line 406? Or is there some chance that the "active chainId" doesn't already have a polling set.

The legacy globally selected polling loop is separate from the PollingController loop(s). So yes, it's very possible that the active chain ID for the globally selected network isn't being polled in the networkClientId polling set

Reply:
done https://github.com/MetaMask/metamask-extension/pull/21868/commits/39e22ac54292816fdc07abd851a4c027a8aef987